torch/utils/data/datapipes/_typing.py

```diff
@@ -422,25 +465,43 @@ def wrap_generator(*args, **kwargs):
                 raise
 
         namespace['__iter__'] = wrap_generator
-    else:
+    else:  # ``__iter__`` of IterDataPipe is NOT a generator function
         # IterDataPipe is an iterator with both ``__iter__`` and ``__next__``
-        if '__next__' in namespace:
+        # And ``__iter__`` may or may not return `self`
+        if '__next__' in namespace:  # If `__next__` exists, put a wrapper around it
             next_func = namespace['__next__']
 
             @functools.wraps(next_func)
             def wrap_next(*args, **kwargs):
                 with context():
+                    datapipe = args[0]
+                    # The `_valid_iterator_id` should either be never set (`None`), or set by at most one
+                    # iterator (`0`). Otherwise, it means there are multiple iterators.
+                    if datapipe._valid_iterator_id is not None and datapipe._valid_iterator_id != 0:
+                        raise RuntimeError(
+                            "A separate iterator has been created for this IterDataPipe, but `next()` has been called"
+                            f"on the original IterDataPipe as well: {_generate_iterdatapipe_msg(datapipe)}\n"
+                            "Multiple iterator references to the same IterDataPipe is not allowed. We recommend "
+                            "using `.fork()` if that is necessary.\n"
+                            "For feedback regarding this single iterator per IterDataPipe constraint, feel free "
+                            "to comment on this issue: https://github.com/pytorch/data/issues/45.")
                     return next_func(*args, **kwargs)
 
             namespace['__next__'] = wrap_next
-        # ``__iter__`` of IterDataPipe returns an iterator other than self
-        else:
-            @functools.wraps(func)
-            def wrap_iter(*args, **kwargs):
-                iter_ret = func(*args, **kwargs)
-                return IteratorDecorator(iter_ret)
 
-            namespace['__iter__'] = wrap_iter
+            # Note that if the `__next__` and `__iter__` do something completely unrelated? It may cause issue but
+            # the user will be violating the iterator protocol
```

Comment:
If the `__next__` and `__iter__` do something completely unrelated, it would violate the iterator protocol and the behavior here maybe unexpected. But as long as user sticks with iterator protocol or returns an iterator, everything should be fine.